### PR TITLE
[Profiler] set_class util (part 1 of Record Optimizer)

### PR DESCRIPTION
**Pull request by slgong-fb**
Summary:
Part 1 of Record Optimizer param_groups and states (https://github.com/pytorch/pytorch/pull/84063)
- nnModule and Optimizer have duplicated parts
- create a util function to avoid duplication

Test Plan: buck run mode/opt //caffe2/test:profiler

Differential Revision: D39397210

### CODE REVIEW

File: torch/csrc/autograd/profiler_python.cpp [hunk 1]
```diff
@@ -249,6 +256,28 @@ class ValueCache {
 
   void trimPrefixes();
 
+  template <CallType C>
+  typename Config<C>::cls_t set_class(
```
Comment (robieta) [inline torch/csrc/autograd/profiler_python.cpp:260]: This feels strange as part of ValueCache. It seems like there should be a base class for the cache of these specializations.

File: torch/csrc/autograd/profiler_python.cpp [hunk 2]
```diff
@@ -249,6 +256,28 @@ class ValueCache {
...
+      typename Config<C>::cache_t& cache,
+      const typename Config<C>::key_t& key,
+      const typename Config<C>::ephemeral_t& frame,
+      const PyCodeObject* code_match) {
```
Comment (robieta) [inline torch/csrc/autograd/profiler_python.cpp:264]: If we template get code on CallType then we can make this more generic. E.g. `TORCH_INTERNAL_ASSERT(code.get() == getCode<C>());`

Comment (slgong-fb) [reply torch/csrc/autograd/profiler_python.cpp:264]: updated

File: torch/csrc/autograd/profiler_python.cpp (hunk 1 above)
Comment (slgong-fb) [reply torch/csrc/autograd/profiler_python.cpp:260]: updated as a standalone function

File: torch/csrc/autograd/profiler_python.cpp [hunk 3]
```diff
@@ -77,6 +77,14 @@ PyCodeObject* nnModuleCode() {
   return module_call_code;
 }
 
+template <CallType C>
+PyCodeObject* getCode();
+
+template <>
+PyCodeObject* getCode<CallType::PyModuleCall>() {
```
Comment (robieta) [inline torch/csrc/autograd/profiler_python.cpp:84]: I don't think we need this indirection. We can just move the body and use `getCode<CallType::PyModuleCall>()` everywhere.

**Review by robieta — APPROVED** [review]
One nit, but overall looks good!

**Pull request closed without merging**